cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -1,636 +1,526 @@
-      store.(InitializeParameterInstruction).getIRVariable()
-    or
-    result.(SourceIRVariableIndirection).getUnderlyingIRVariable() =
-      store.(InitializeIndirectionInstruction).getIRVariable()
+  override Cpp::Location getLocation() { result = this.getDefiningInstruction().getLocation() }
+
+  final override predicate hasIndexInBlock(IRBlock block, int index) {
+    this.getDefiningInstruction() = block.getInstruction(index)
   }
 
-  override predicate isCertain() { any() }
+  predicate isCertain() { isDef(true, _, address, _, _, ind) }
 }
 
-abstract class Use extends DefOrUse {
-  Operand use;
+class UseImpl extends DefOrUseImpl, TUseImpl {
+  Operand operand;
+  int ind;
 
-  override Operand asUse() { result = use }
+  UseImpl() { this = TUseImpl(operand, ind) }
 
-  /** Gets the underlying operand of this use. */
-  Operand getOperand() { result = use }
+  Operand getOperand() { result = operand }
 
-  override string toString() { result = "Use" }
+  override string toString() { result = "UseImpl" }
 
-  /** Gets the variable that is used by this use. */
-  abstract SourceVariable getSourceVariable();
+  final override predicate hasIndexInBlock(IRBlock block, int index) {
+    operand.getUse() = block.getInstruction(index)
+  }
 
-  override IRBlock getBlock() { result = use.getUse().getBlock() }
+  final override IRBlock getBlock() { result = operand.getUse().getBlock() }
 
-  override Cpp::Location getLocation() { result = use.getLocation() }
-}
+  final override Cpp::Location getLocation() { result = operand.getLocation() }
 
-private class ExplicitUse extends Use, TExplicitUse {
-  ExplicitUse() { this = TExplicitUse(use) }
+  final int getIndirection() { isUse(_, operand, _, result, ind) }
 
-  override SourceVariable getSourceVariable() {
-    exists(VariableInstruction var |
-      use.getDef() = var and
-      if use.getUse() instanceof ReadSideEffectInstruction
-      then result.(SourceIRVariableIndirection).getUnderlyingIRVariable() = var.getIRVariable()
-      else result.(SourceIRVariable).getIRVariable() = var.getIRVariable()
-    )
-  }
-}
+  override int getIndirectionIndex() { result = ind }
+
+  override Instruction getBase() { isUse(_, operand, result, _, ind) }
 
-private class ReturnParameterIndirection extends Use, TReturnParamIndirection {
-  ReturnParameterIndirection() { this = TReturnParamIndirection(use) }
+  predicate isCertain() { isUse(true, operand, _, _, ind) }
+}
 
-  override SourceVariable getSourceVariable() {
-    exists(ReturnIndirectionInstruction ret |
-      returnParameterIndirection(use, ret) and
-      result.(SourceIRVariableIndirection).getUnderlyingIRVariable() = ret.getIRVariable()
+predicate adjacentDefRead(DefOrUse defOrUse1, UseOrPhi use) {
+  exists(IRBlock bb1, int i1, SourceVariable v |
+    defOrUse1.asDefOrUse().hasIndexInBlock(bb1, i1, v)
+  |
+    exists(IRBlock bb2, int i2 |
+      adjacentDefRead(_, pragma[only_bind_into](bb1), pragma[only_bind_into](i1),
```

Comment:
I _think_ this could in principle go wrong, if `adjacentDefRead(def, bb1, i1, bb2, i2)` happens to hold for a _different_ SSA definition (with another source variable) than the one underlying `defOrUse1`. But only if multiple variables could be referenced at the same index (which it can in C# for psuedo-reads).

Reply:
Isn't this okay because we're also requiring that:
```
defOrUse1.asDefOrUse().hasIndexInBlock(bb1, i1, v) and
use.asDefOrUse().(UseImpl).hasIndexInBlock(bb2, i2, v)
```
(i.e., that they read the same source variable.)

Would you be more confident in the correctness if the `Ssa::Definition` column wasn't ignored, and instead checked the source variable of that definition?

Reply:
Ah yes, for some reason I failed to spot the `v`, so all should be good.